File: image_cache.py
"""Image caching functionality for the roxie deployment tool."""

import json
import os
import subprocess

from rich.console import Console

from logger import Logger


class ImageCache:
    """Manages cache of verified pullable Docker images"""

    def __init__(self, logger: Logger, cache_file: str | None = None, max_entries: int = 20):
        """Initialize ImageCache with optional cache file path and max entries"""
        self.cache_file = cache_file or os.path.expanduser("~/.roxie.image_cache")
        self.max_entries = max_entries
        self._cache = self.load_cache()
        self.logger = logger

    def load_cache(self) -> list[str]:
        """Load image cache from file"""
        try:
            if os.path.exists(self.cache_file):
                with open(self.cache_file) as f:
                    data = json.load(f)
                    # Handle both old format (list) and new format (dict with 'images' key)
                    if isinstance(data, list):
                        return data
                    elif isinstance(data, dict):
                        images_value = data.get("images", [])
                        # Ensure we always return a list[str]
                        return [str(x) for x in images_value]
        except (json.JSONDecodeError, OSError):
            pass
        return []

    def save_cache(self):
        """Save image cache to file"""
        try:
            # Ensure directory exists
            os.makedirs(os.path.dirname(self.cache_file), exist_ok=True)

            with open(self.cache_file, "w") as f:
                json.dump({"images": self._cache}, f, indent=2)
        except OSError:
            # Silently fail on cache save errors
            pass

    def is_cached(self, image_ref: str) -> bool:
        """Check if image is in cache"""
        return image_ref in self._cache

    def add_to_cache(self, image_ref: str):
        """Add image to cache, maintaining max_entries limit"""
        if image_ref in self._cache:
            # Move to end (most recent)
            self._cache.remove(image_ref)

        self._cache.append(image_ref)

        # Maintain max entries
        if len(self._cache) > self.max_entries:
            self._cache = self._cache[-self.max_entries :]

        self.save_cache()

    def test_cache_functionality(self, console: Console):
        """Test cache read/write functionality"""
        try:
            # Test write
            test_image = "test/image:latest"
            original_cache = self._cache.copy()
            self.add_to_cache(test_image)

            # Test read
            if self.is_cached(test_image):
                console.print("✓ Cache test passed", style="dim green")
            else:
                console.print("⚠ Cache test failed", style="dim yellow")

            # Restore original cache
            self._cache = original_cache
            self.save_cache()
        except Exception as e:
            console.print(f"Cache test failed: {str(e)}", style="dim yellow")

    def verify_image_pullable(self, image_ref: str) -> bool:
        """Verify if image is pullable, using cache when possible"""
        if self.is_cached(image_ref):
            return True

        # Use skopeo to check if image is pullable
        try:
            result = subprocess.run(
                ["skopeo", "inspect", "--raw", f"docker://{image_ref}"], capture_output=True, text=True, timeout=30
            )

            if result.returncode == 0:
                self.add_to_cache(image_ref)
                return True
            else:
                print(result.stderr)
                print(result.stdout)

        except (subprocess.TimeoutExpired, subprocess.CalledProcessError, FileNotFoundError):
            pass

        return False

    def verify_images_pullable(self, *images: str) -> bool:
        """Verify that Docker images are pullable using cached results for speed"""
        if not images:
            return True

        # Skip verification if environment variable is set
        if os.environ.get("SKIP_IMAGE_VERIFICATION", "").lower() in ("true", "1", "yes"):
            self.logger.print_with_timestamp(
                f"Skipping image verification for {len(images)} images (SKIP_IMAGE_VERIFICATION=true)"
            )
            return True

        # Separate cached and uncached images
        cached_images = []
        uncached_images = []

        for img in images:
            if self.is_cached(img):
                cached_images.append(img)
            else:
                uncached_images.append(img)

        # Report cached results immediately
        if cached_images:
            self.logger.print_with_timestamp(f"✓ {len(cached_images)} images verified from cache", style="dim green")
            for img in cached_images:
                self.logger.print_with_timestamp(f"✓ Image {img} (cached)", style="dim green")

        # Verify uncached images if any
        failed_images = []
        if uncached_images:
            import concurrent.futures

            def verify_single_image(img: str) -> tuple[str, bool, str]:
                """Verify a single image using ImageCache"""
                if self.verify_image_pullable(img):
                    return (img, True, "")
                else:
                    return (img, False, "not pullable")

            # Verify uncached images in parallel (max 4 concurrent to avoid overwhelming registry)
            max_workers = min(4, len(uncached_images))

            with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
                # Submit all verification tasks
                future_to_image = {executor.submit(verify_single_image, img): img for img in uncached_images}

                # Collect results as they complete
                for future in concurrent.futures.as_completed(future_to_image):
                    img, success, error_msg = future.result()

                    if success:
                        self.logger.print_with_timestamp(f"✓ Image {img} verified", style="dim green")
                    else:
                        self.logger.print_with_timestamp(f"✗ Image {img} failed: {error_msg}", style="bold red")
                        failed_images.append((img, error_msg))

        if failed_images:
            self.logger.error(f"Failed to verify {len(failed_images)} images:")
            for img, error_msg in failed_images:
                self.logger.error(f"  - {img}: {error_msg}")
            return False

        total_images = len(images)
        cached_count = len(cached_images)
        verified_count = len(uncached_images) - len(failed_images)

        self.logger.print_with_timestamp(
            f"✓ All {total_images} images verified successfully ({cached_count} cached, {verified_count} verified)",
            style="bold green",
        )
        return True

File: internal/imagecache/imagecache.go
package imagecache

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"github.com/stackrox/roxie/internal/logger"
	"github.com/stackrox/roxie/internal/skopeohelper"
)

// ImageCache manages cache of verified pullable Docker images
type ImageCache struct {
	cacheFile  string
	maxEntries int
	cache      []string
	logger     *logger.Logger
	mu         sync.Mutex
}

// CacheData represents the structure of the cache file
type CacheData struct {
	Images []string `json:"images"`
}

// New creates a new ImageCache instance
func New(log *logger.Logger, cacheFile string, maxEntries int) *ImageCache {
	if cacheFile == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			home = "."
		}
		cacheFile = filepath.Join(home, ".roxie.image_cache")
	}

	if maxEntries <= 0 {
		maxEntries = 20
	}

	ic := &ImageCache{
		cacheFile:  cacheFile,
		maxEntries: maxEntries,
		logger:     log,
	}

	ic.cache = ic.loadCache()
	return ic
}

// loadCache loads image cache from file
func (ic *ImageCache) loadCache() []string {
	data, err := os.ReadFile(ic.cacheFile)
	if err != nil {
		return []string{}
	}

	// Try new format first (with "images" key)
	var cacheData CacheData
	if err := json.Unmarshal(data, &cacheData); err == nil && cacheData.Images != nil {
		return cacheData.Images
	}

	// Try old format (plain array)
	var images []string
	if err := json.Unmarshal(data, &images); err == nil {
		return images
	}

	return []string{}
}

// saveCache saves image cache to file
// Must be called with ic.mu held
func (ic *ImageCache) saveCache() {
	// Ensure directory exists
	dir := filepath.Dir(ic.cacheFile)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return // Silently fail on cache save errors
	}

	cacheData := CacheData{Images: ic.cache}
	data, err := json.MarshalIndent(cacheData, "", "  ")
	if err != nil {
		return
	}

	_ = os.WriteFile(ic.cacheFile, data, 0644)
}

// IsCached checks if image is in cache
func (ic *ImageCache) IsCached(imageRef string) bool {
	ic.mu.Lock()
	defer ic.mu.Unlock()

	for _, img := range ic.cache {
		if img == imageRef {
			return true
		}
	}
	return false
}

// AddToCache adds image to cache, maintaining max_entries limit
func (ic *ImageCache) AddToCache(imageRef string) {
	ic.mu.Lock()
	defer ic.mu.Unlock()

	// Remove if already exists (to move to end)
	for i, img := range ic.cache {
		if img == imageRef {
			ic.cache = append(ic.cache[:i], ic.cache[i+1:]...)
			break
		}
	}

	// Add to end
	ic.cache = append(ic.cache, imageRef)

	// Maintain max entries
	if len(ic.cache) > ic.maxEntries {
		ic.cache = ic.cache[len(ic.cache)-ic.maxEntries:]
	}

	ic.saveCache()
}

// VerifyImagePullable verifies if image is pullable using cache when possible
func (ic *ImageCache) VerifyImagePullable(ctx context.Context, imageRef string) bool {
	if ic.IsCached(imageRef) {
		return true
	}

	// Use OCI registry client to verify image accessibility.
	err := skopeohelper.InspectImage(ctx, ic.logger, imageRef)
	if err == nil {
		ic.AddToCache(imageRef)
		return true
	}

	fmt.Fprintf(os.Stderr, "Failed to verify image %s: %v\n", imageRef, err)
	return false
}

// VerifyImagesPullable verifies that Docker images are pullable using cached results
func (ic *ImageCache) VerifyImagesPullable(ctx context.Context, images ...string) bool {
	if len(images) == 0 {
		return true
	}

	// Skip verification if environment variable is set
	if skip := os.Getenv("SKIP_IMAGE_VERIFICATION"); skip == "true" || skip == "1" || skip == "yes" {
		ic.logger.Infof("Skipping image verification for %d images (SKIP_IMAGE_VERIFICATION=true)", len(images))
		return true
	}

	// Separate cached and uncached images
	var cachedImages, uncachedImages []string
	for _, img := range images {
		if ic.IsCached(img) {
			cachedImages = append(cachedImages, img)
		} else {
			uncachedImages = append(uncachedImages, img)
		}
	}

	// Report cached results immediately
	if len(cachedImages) > 0 {
		ic.logger.Successf("✓ %d images verified from cache", len(cachedImages))
		for _, img := range cachedImages {
			ic.logger.Dim(fmt.Sprintf("✓ Image %s (cached)", img))
		}
	}

	// Verify uncached images if any
	failedImages := []string{}
	if len(uncachedImages) > 0 {
		// In Go, we can use goroutines for parallel verification
		type result struct {
			img     string
			success bool
			errMsg  string
		}

		results := make(chan result, len(uncachedImages))
		maxWorkers := 4
		if len(uncachedImages) < maxWorkers {
			maxWorkers = len(uncachedImages)
		}

		// Worker pool
		sem := make(chan struct{}, maxWorkers)
		var wg sync.WaitGroup

		for _, img := range uncachedImages {
			wg.Add(1)
			go func(image string) {
				defer wg.Done()
				sem <- struct{}{}        // Acquire semaphore
				defer func() { <-sem }() // Release semaphore

				success := ic.VerifyImagePullable(ctx, image)
				if success {
					results <- result{img: image, success: true}
				} else {
					results <- result{img: image, success: false, errMsg: "not pullable"}
				}
			}(img)
		}

		// Close results channel when all workers are done
		go func() {
			wg.Wait()
			close(results)
		}()

		// Collect results
		for res := range results {
			if res.success {
				ic.logger.Dim(fmt.Sprintf("✓ Image %s verified", res.img))
			} else {
				ic.logger.Errorf("✗ Image %s failed: %s", res.img, res.errMsg)
				failedImages = append(failedImages, res.img)
			}
		}
	}

	if len(failedImages) > 0 {
		ic.logger.Errorf("Failed to verify %d images:", len(failedImages))
		for _, img := range failedImages {
			ic.logger.Errorf("  - %s", img)
		}
		return false
	}

	cachedCount := len(cachedImages)
	verifiedCount := len(uncachedImages) - len(failedImages)
	ic.logger.Successf("✓ All %d images verified successfully (%d cached, %d verified)",
		len(images), cachedCount, verifiedCount)

	return true
}
